node/src/dev_service.rs

```diff
@@ -0,0 +1,264 @@
+		inherent_data_providers,
+		other: (block_import, pending_transactions),
+	} = new_partial(&config, author_id)?;
+
+	let (network, network_status_sinks, system_rpc_tx, network_starter) =
+		sc_service::build_network(sc_service::BuildNetworkParams {
+			config: &config,
+			client: client.clone(),
+			transaction_pool: transaction_pool.clone(),
+			spawn_handle: task_manager.spawn_handle(),
+			import_queue,
+			on_demand: None,
+			block_announce_validator_builder: None,
+		})?;
+
+	// Channel for the rpc handler to communicate with the authorship task.
+	let (command_sink, commands_stream) = futures::channel::mpsc::channel(1000);
+
+	if config.offchain_worker.enabled {
+		sc_service::build_offchain_workers(
+			&config,
+			backend.clone(),
+			task_manager.spawn_handle(),
+			client.clone(),
+			network.clone(),
+		);
+	}
+
+	let role = config.role.clone();
+	let prometheus_registry = config.prometheus_registry().cloned();
+	let telemetry_connection_sinks = sc_service::TelemetryConnectionSinks::default();
+	let is_authority = role.is_authority();
+	let subscription_task_executor =
+		sc_rpc::SubscriptionTaskExecutor::new(task_manager.spawn_handle());
+
+	let rpc_extensions_builder = {
+		let client = client.clone();
+		let pool = transaction_pool.clone();
+		let network = network.clone();
+		let pending = pending_transactions.clone();
+		Box::new(move |deny_unsafe, _| {
+			let deps = crate::rpc::FullDeps {
+				client: client.clone(),
+				pool: pool.clone(),
+				graph: pool.pool().clone(),
+				deny_unsafe,
+				is_authority,
+				network: network.clone(),
+				pending_transactions: pending.clone(),
+				command_sink: Some(command_sink.clone()),
+			};
+			crate::rpc::create_full(deps, subscription_task_executor.clone())
+		})
+	};
+
+	sc_service::spawn_tasks(sc_service::SpawnTasksParams {
+		network: network.clone(),
+		client: client.clone(),
+		keystore: keystore_container.sync_keystore(),
+		task_manager: &mut task_manager,
+		transaction_pool: transaction_pool.clone(),
+		telemetry_connection_sinks,
+		rpc_extensions_builder,
+		on_demand: None,
+		remote_blockchain: None,
+		backend,
+		network_status_sinks,
+		system_rpc_tx,
+		config,
+	})?;
+
+	// Spawn Frontier pending transactions maintenance task (as essential, otherwise we leak).
+	if pending_transactions.is_some() {
+		use fp_consensus::{ConsensusLog, FRONTIER_ENGINE_ID};
+		use futures::StreamExt;
+		use sp_runtime::generic::OpaqueDigestItemId;
+
+		const TRANSACTION_RETAIN_THRESHOLD: u64 = 5;
+		task_manager.spawn_essential_handle().spawn(
+			"frontier-pending-transactions",
+			client
+				.import_notification_stream()
+				.for_each(move |notification| {
+					if let Ok(locked) = &mut pending_transactions.clone().unwrap().lock() {
+						// As pending transactions have a finite lifespan anyway
+						// we can ignore MultiplePostRuntimeLogs error checks.
+						let mut frontier_log: Option<_> = None;
+						for log in notification.header.digest.logs {
```

Comment:
So, this loop sets `frontier_log` to the last log that `is_some`  in `notification.header.digest.logs`. If there are no logs that are `is_some`, then it sets it to None. Why start from the beginning? Shouldn't we just start from the end and go backwards and set `frontier_log` to the first `log` that `is_some`. The worst case is still the same if they are all None, but the best case is much better.

so like
```rust
for log in notification.header.digest.logs.iter().rev() {
    let log = log.try_to....();
    if log.is_some() {
        frontier_log = log;
        break;
    }
}
```

Reply:
This is a great point. But it is not introduced in this PR. It only seems so because I copied the dev service into a new place. You can see the same logic previously in the [standalone](https://github.com/PureStake/moonbeam/blob/master/node/standalone/src/service.rs#L310) and [parachain](https://github.com/PureStake/moonbeam/blob/master/node/parachain/src/service.rs#L249) nodes.

I've created a followup issue for this https://github.com/PureStake/moonbeam/issues/217